akka23-conductr-client-lib/src/main/scala/com/typesafe/conductr/clientlib/akka/BundlesConnector.scala

```diff
@@ -81,56 +165,111 @@ object BundlesConnector {
-  ): Unit =
-    for {
-      eventStreamResult <- getBundlesEvents()
-    } eventStreamResult match {
-      case EventStreamSuccess(sseSource) =>
-        sseSource
-          .runForeach(connector ! _)
-          .map(done => connector ! ConsumeBundlesEventsOutcome(Success(done)))
-          .recover {
-            case error =>
-              connector ! ConsumeBundlesEventsOutcome(Failure(error))
+  private def consumeBundlesEvents(connector: ActorRef, connection: Connection, apiVersion: ApiVersion.Value)(implicit ec: ExecutionContext, mat: ActorMaterializer): Unit = {
+    import EventStreamUnmarshalling._
+
+    Source
+      .single(bundlesEventsRequest(apiVersion))
+      .via(connection)
+      .alsoTo(
+        Sink.head.mapMaterializedValue(_.foreach { response =>
+          if (response.status.isSuccess())
+            connector ! GetBundles
+        })
+      )
+      .mapAsync(1) { response =>
+        Unmarshal(response.entity.withSizeLimit(Long.MaxValue))
+          .to[Source[ServerSentEvent, Any]]
+          .flatMap { sseSource =>
+            sseSource
+              .filter(_.data != "")
+              .runForeach(connector ! _)
           }
+      }
+      .runWith(
+        Sink.onComplete {
+          case outcome => connector ! ConsumeBundlesEventsOutcome(outcome)
+        }
+      )
+  }
+
+  private def bundlesEventsRequest(apiVersion: ApiVersion.Value): HttpRequest =
+    Get(s"${versionPrefix(apiVersion)}/bundles/events")
+      .addHeader(`Cache-Control`(`no-cache`))
 
-      case EventStreamFailure(httpResponseCode, errorMessage) =>
-        connector ! ConsumeBundlesEventsOutcome(
-          Failure(new Exception(s"HTTP failure when obtaining bundle events - code [$httpResponseCode] - error [$errorMessage]"))
-        )
-    }
+  private def versionPrefix(apiVersion: ApiVersion.Value): String =
+    s"/v$apiVersion"
 }
 
 /**
- * Connects to the /bundles endpoint of ConductR and triggers the given [[callback]] when updates are available.
+ * Connects to the `/bundles` endpoint of ConductR and triggers the given [[connectUntil]] function when updates from
+ * the `/bundles/events` SSE are available.
+ *
+ * If [[connectUntil]] returns `true`, [[BundlesConnector]] will terminate itself, otherwise [[BundlesConnector]] will
+ * continue receiving bundle SSE until the next update is available.
+ *
+ * Upon startup, [[BundlesConnector]] will obtain the bundle state from the `/bundles` endpoint, and apply the
+ * [[connectUntil]] function to the obtained bundle state. If the [[connectUntil]] condition is not satisfied,
+ * [[BundlesConnector]] will proceed to connect to the `/bundle/events` endpoint of ConductR.
  */
-class BundlesConnector[T](
-    getBundlesEvents: BundlesConnector.GetBundleEvents,
-    getBundlesInfo: BundlesConnector.GetBundlesInfo,
+class BundlesConnector(
+    conductrAddress: URL,
+    apiVersion: ApiVersion.Value,
+    connectUntil: Seq[Bundle] => Boolean,
```

Comment:
How about instead of `connectUntil` we consider `foreach` along with some separate connection object. `foreach` conveys that side-effects are permitted. As the params to `foreach`, pass some control object that can be "told" to discontinue further processing, along with the sequence of bundles. The end result in client code could be something like:

```scala
val (_, result) = 
  BundlesConnector.connect(conductrAddress, apiVersion, stopAfter = completeTimeout)
    .foreach {
      case (control, bundles) if runningBundlesCount(bundles) >= requiredScale => control.stop()
      case _ =>
    }
```

wdyt?